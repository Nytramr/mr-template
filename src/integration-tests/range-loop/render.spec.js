import { Template } from '../../index';

describe('Template with Range Loop', () => {
  test('simple range', () => {
    const rangeLoopTpl =
      'Beginning of the file' +
      '\n' +
      '{{range .range}}' +
      'Loop {{.}}, ' +
      '{{end}}' +
      '\n' +
      'End of the file' +
      '\n';

    const template = new Template('My template');
    expect(() => template.parse(rangeLoopTpl)).not.toThrow();

    expect(template.execute({ range: [1, 2, 3, 4, 5] })).toMatchSnapshot();
    expect(
      template.execute({ range: { a: 1, b: 2, c: 3, d: 4, e: 5, f: 6 } }),
    ).toMatchSnapshot();
    expect(template.execute({ range: 'abcdef' })).toMatchSnapshot();
  });

  test('nested range', () => {
    const rangeLoopTpl = `Beginning of the file
{{- range .range}}
  Loop:
  {{- range . -}}
    {{' '}}{{.}},
  {{- end}}
  End loop
{{end -}}
End of the file`;

    const template = new Template('My template');
    expect(() => template.parse(rangeLoopTpl)).not.toThrow();

    expect(
      template.execute({ range: [[1, 2, 3], [4], [5, 6]] }),
    ).toMatchSnapshot();
    expect(
      template.execute({ range: { a: 'abc', b: ['d', 'e'], c: 'f', d: 'g' } }),
    ).toMatchSnapshot();
  });

  test('range with a constant', () => {
    const rangeLoopTpl =
      'Beginning of the file' +
      '\n' +
      '{{range "str range render"}}' +
      '{{.}}, ' +
      '{{end}}' +
      '\n' +
      'End of the file' +
      '\n';

    const template = new Template('My template');
    expect(() => template.parse(rangeLoopTpl)).not.toThrow();

    const render = template.execute({});

    expect(render).toMatchSnapshot();
    expect(template.execute('Some other value')).toEqual(render);
  });

  test('with "else" fallback', () => {
    const rangeLoopTpl =
      'Beginning of the file' +
      '\n' +
      '{{range .range}}' +
      'Loop {{.}}' +
      '\n' +
      '{{else}}' +
      '\n' +
      'The range is empty' +
      '\n' +
      '{{end}}' +
      '\n' +
      'End of the file' +
      '\n';

    const template = new Template('My template');
    expect(() => template.parse(rangeLoopTpl)).not.toThrow();

    expect(template.execute({ range: 5 })).toMatchSnapshot();
    expect(template.execute({ range: 0 })).toMatchSnapshot();
  });

  test('with "break"', () => {
    const rangeLoopTpl =
      'Beginning of the file\n' +
      '{{range .range}}' +
      'Loop {{.}}\n' +
      '{{if eq . 3}}' +
      '{{break}}' +
      '{{end}}' +
      '{{end}}' +
      'End of the file\n';

    const template = new Template('My template');
    expect(() => template.parse(rangeLoopTpl)).not.toThrow();

    expect(template.execute({ range: 10 })).toMatchSnapshot();
    expect(
      template.execute({ range: [1, '5', 'a', 30, '3', 'b', 3, 'c'] }),
    ).toMatchSnapshot();
  });

  test('with "continue"', () => {
    const rangeLoopTpl =
      'Beginning of the file\n' +
      '{{range .range}}' +
      '{{if gt . 3}}' +
      '{{continue}}' +
      '{{end}}' +
      'Loop {{.}}\n' +
      '{{end}}' +
      'End of the file\n';

    const template = new Template('My template');
    expect(() => template.parse(rangeLoopTpl)).not.toThrow();

    expect(template.execute({ range: 10 })).toMatchSnapshot();
    expect(
      template.execute({ range: [1, 5, 'a', 30, 6, 'b', 3, 'c'] }),
    ).toMatchSnapshot();
  });

  describe('With variables', () => {
    describe('declared inline', () => {
      test('element only', () => {
        const rangeLoopTpl =
          'Beginning of the file' +
          '\n' +
          '{{range $x := .range}}' +
          'Loop {{$x}},' +
          '{{end}}' +
          '\n' +
          'End of the file' +
          '\n';

        const template = new Template('My template');
        expect(() => template.parse(rangeLoopTpl)).not.toThrow();

        expect(template.execute({ range: [1, 2, 3, 4, 5] })).toMatchSnapshot();
        expect(template.execute({ range: 'abcdef' })).toMatchSnapshot();
      });

      test('key and element', () => {
        const rangeLoopTpl =
          'Beginning of the file' +
          '\n' +
          '{{range $i, $x := .range}}' +
          'Loop {{$i}}: {{$x}},' +
          '{{end}}' +
          '\n' +
          'End of the file' +
          '\n';

        const template = new Template('My template');
        expect(() => template.parse(rangeLoopTpl)).not.toThrow();

        expect(
          template.execute({ range: ['a', 'b', 'c', 'd', 'e', 'f'] }),
        ).toMatchSnapshot();
        expect(
          template.execute({ range: { a: 1, b: 2, c: 3, d: 4, e: 5, f: 6 } }),
        ).toMatchSnapshot();
      });
    });

    describe('previously declared', () => {
      test('element only', () => {
        const rangeLoopTpl =
          'Beginning of the file' +
          '\n' +
          '{{$x := ""}}' +
          '{{range $x = .range}}' +
          'Loop {{$x}},' +
          '{{end}}' +
          '\n' +
          'End of the file' +
          '\n';

        const template = new Template('My template');
        expect(() => template.parse(rangeLoopTpl)).not.toThrow();

        expect(template.execute({ range: [1, 2, 3, 4] })).toMatchSnapshot();
        expect(template.execute({ range: 'abcde' })).toMatchSnapshot();
      });

      test('key and element', () => {
        const rangeLoopTpl =
          'Beginning of the file' +
          '\n' +
          '{{$i := ""}}' +
          '{{$x := ""}}' +
          '{{range $i, $x = .range}}' +
          'Loop {{$i}}: {{$x}},' +
          '{{end}}' +
          '\n' +
          'End of the file' +
          '\n';

        const template = new Template('My template');
        expect(() => template.parse(rangeLoopTpl)).not.toThrow();

        expect(
          template.execute({ range: ['A', 'B', 'C', 'D'] }),
        ).toMatchSnapshot();
        expect(
          template.execute({ range: { a: 'X', b: 'Y', c: 'Z' } }),
        ).toMatchSnapshot();
      });
    });
  });
});
