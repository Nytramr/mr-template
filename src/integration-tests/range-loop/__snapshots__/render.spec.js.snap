// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template with Range Loop With variables declared inline element only 1`] = `
"Beginning of the file
Loop 1,Loop 2,Loop 3,Loop 4,Loop 5,
End of the file
"
`;

exports[`Template with Range Loop With variables declared inline element only 2`] = `
"Beginning of the file
Loop a,Loop b,Loop c,Loop d,Loop e,Loop f,
End of the file
"
`;

exports[`Template with Range Loop With variables declared inline key and element 1`] = `
"Beginning of the file
Loop 0: a,Loop 1: b,Loop 2: c,Loop 3: d,Loop 4: e,Loop 5: f,
End of the file
"
`;

exports[`Template with Range Loop With variables declared inline key and element 2`] = `
"Beginning of the file
Loop a: 1,Loop b: 2,Loop c: 3,Loop d: 4,Loop e: 5,Loop f: 6,
End of the file
"
`;

exports[`Template with Range Loop With variables previously declared element only 1`] = `
"Beginning of the file
Loop 1,Loop 2,Loop 3,Loop 4,
End of the file
"
`;

exports[`Template with Range Loop With variables previously declared element only 2`] = `
"Beginning of the file
Loop a,Loop b,Loop c,Loop d,Loop e,
End of the file
"
`;

exports[`Template with Range Loop With variables previously declared key and element 1`] = `
"Beginning of the file
Loop 0: A,Loop 1: B,Loop 2: C,Loop 3: D,
End of the file
"
`;

exports[`Template with Range Loop With variables previously declared key and element 2`] = `
"Beginning of the file
Loop a: X,Loop b: Y,Loop c: Z,
End of the file
"
`;

exports[`Template with Range Loop nested range 1`] = `
"Beginning of the file
  Loop: 1, 2, 3,
  End loop

  Loop: 4,
  End loop

  Loop: 5, 6,
  End loop
End of the file"
`;

exports[`Template with Range Loop nested range 2`] = `
"Beginning of the file
  Loop: a, b, c,
  End loop

  Loop: d, e,
  End loop

  Loop: f,
  End loop

  Loop: g,
  End loop
End of the file"
`;

exports[`Template with Range Loop range with a constant 1`] = `
"Beginning of the file
s, t, r,  , r, a, n, g, e,  , r, e, n, d, e, r, 
End of the file
"
`;

exports[`Template with Range Loop simple range 1`] = `
"Beginning of the file
Loop 1, Loop 2, Loop 3, Loop 4, Loop 5, 
End of the file
"
`;

exports[`Template with Range Loop simple range 2`] = `
"Beginning of the file
Loop 1, Loop 2, Loop 3, Loop 4, Loop 5, Loop 6, 
End of the file
"
`;

exports[`Template with Range Loop simple range 3`] = `
"Beginning of the file
Loop a, Loop b, Loop c, Loop d, Loop e, Loop f, 
End of the file
"
`;

exports[`Template with Range Loop with "break" 1`] = `
"Beginning of the file
Loop 0
Loop 1
Loop 2
Loop 3
End of the file
"
`;

exports[`Template with Range Loop with "break" 2`] = `
"Beginning of the file
Loop 1
Loop 5
Loop a
Loop 30
Loop 3
Loop b
Loop 3
End of the file
"
`;

exports[`Template with Range Loop with "continue" 1`] = `
"Beginning of the file
Loop 0
Loop 1
Loop 2
Loop 3
"
`;

exports[`Template with Range Loop with "continue" 2`] = `
"Beginning of the file
Loop 1
Loop a
Loop b
Loop 3
Loop c
"
`;

exports[`Template with Range Loop with "else" fallback 1`] = `
"Beginning of the file
Loop 0
Loop 1
Loop 2
Loop 3
Loop 4

End of the file
"
`;

exports[`Template with Range Loop with "else" fallback 2`] = `
"Beginning of the file

The range is empty

End of the file
"
`;
